modules/http4s/test/src/smithy4s/http4s/NullsAndDefaultEncodingSuite.scala

```diff
@@ -0,0 +1,262 @@
+          requiredHeaderWithDefault = "required-header-with-default-override"
+        )
+      )
+    IO.pure(
+      assert.same(
+        Document.obj(
+          "optional" -> Document.fromString("optional-override"),
+          "optionalWithDefault" -> Document.fromString(
+            "optional-default-override"
+          ),
+          "requiredWithDefault" -> Document.fromString(
+            "required-default-override"
+          ),
+          "optionalHeader" -> Document.fromString(
+            "optional-header-override"
+          ),
+          "optionalHeaderWithDefault" -> Document.fromString(
+            "optional-header-with-default-override"
+          ),
+          "requiredHeaderWithDefault" -> Document.fromString(
+            "required-header-with-default-override"
+          )
+        ),
+        result
+      )
+    )
+  }
+
+  object Impl extends ServiceWithNullsAndDefaults[IO] {
+    override def operation(input: OperationInput): IO[OperationOutput] =
+      IO.pure(OperationOutput())
+  }
+
+  private val specHeaders = Set(
+    ci"optional-header",
+    ci"optional-header-with-default",
+    ci"required-header-with-default"
+  )
+
+  case class TestResponse(headers: Map[CIString, String], body: Json)
+
+  case class TestRequest(
+      headers: Map[CIString, String],
+      query: Map[String, String],
+      labels: List[String],
+      body: Json
+  )
+
+  private def runServerTest(explicitDefaults: Boolean): IO[TestResponse] = {
+    def run(
+        routes: HttpRoutes[IO],
+        req: Request[IO]
+    ): IO[(Map[CIString, String], Json)] =
+      routes.orNotFound.run(req).flatMap { response =>
+        response.as[Json].map(headersToMap(response.headers) -> _)
+      }
+    SimpleRestJsonBuilder
+      .withExplicitDefaultsEncoding(explicitDefaults)
+      .routes(Impl)
+      .resource
+      .use { routes =>
+        for {
+          result <- run(
+            routes,
+            Request[IO](method = Method.POST, uri = uri"/operation/label")
+          )
+          (headers, body) = result
+        } yield TestResponse(headers, body)
+      }
+  }
+
+  private def runClientTest(
+      explicitDefaults: Boolean,
+      input: OperationInput
+  ): IO[TestRequest] = {
+    val resources = for {
+      promise <- Deferred[IO, Request[IO]].toResource
+      httpClient: Client[IO] = Client(req =>
+        promise.complete(req).as(Response[IO]()).toResource
+      )
+      client <- SimpleRestJsonBuilder
+        .withExplicitDefaultsEncoding(explicitDefaults)
+        .apply(ServiceWithNullsAndDefaults)
+        .client(httpClient)
+        .resource
+    } yield (promise, client)
+    resources.use { case (promise, client) =>
+      client.operation(input) >> promise.get.flatMap { req =>
+        val query =
+          req.uri.query.pairs.flatMap(kv => kv._2.map(kv._1 -> _)).toMap
+        val labels = req.uri.path.segments
+          .map(_.toString)
+          .filterNot(_ == "operation")
```

Comment:
**question:** why don't we keep the `operation` segment?

Reply:
Up to you, I just found it a bit strange. If it's not adding a lot to the diff, I'd assert on the whole thing.

Reply:
Sure, I don't mind, updated